### ISSUE TITLE
Spawning a background task

### ISSUE THREAD

**Issue by forkbomb9** [async-rs/async-std#817]
I want to spawn a task that handles `TcpStream`s, so that I can handle Ctrl+C correctly. But if I use task::spawn to spawn the TcpListener handler, nothing is run. Unless I `await` the task, but then how do I handle ctrl+c?
If I use std's `thread::spawn` and then `block_on` the async handling function, everything works as expected.

Check out this code (`async-std = "1.6"` in Cargo.toml):

```rust
use async_std::task;
use std::thread;

fn main() {
    // Running the async main
    task::block_on(async_main());
}

async fn async_main() {
    // Setting up the TcpListener, other work
    // ...

    // Using this, "Receiving connections" gets printed
    // thread::spawn(move || {
    //     task::block_on(connection_handler());
    // });

    // Nothing gets printed
    task::spawn(async move {
        connection_handler().await;
    });

    // Here we setup a handler for Ctrl+C

    // Waiting until Ctrl+C is pressed
    loop {}
}

async fn connection_handler() {
    println!("Receiving connections");
    // Eternal loop handling TcpStreams
    loop {}
}
```

**Comment by zhaxzhax** [comment]
The future will do nothing unless you await it. In that case, if you want to hold a thread that loop handling TcpStreams, spawn a thread that wait for a task is a choice. Async is a way to get a result asynchronously, not a way to hold another thread.

**Comment by forkbomb9** [comment]
And not a way to do tasks asynchronously? Because in fact inside of `connection_handler` I'm spawning a new task with a handler that receives `TcpStream`s via a `channel`. And inside of the handler, I'm spawning _another_ task to actually handle the connection. That's why I found strange that I can't spawn that first function...

```rust
async fn connection_handler(server: TcpListener) {
    let mut incoming = server.incoming();

    let (tx, rx) = channel(3);

    let handler = Handler::new(rx);

    task::spawn(async move {
        handler.start().await;
    });

    while let Some(Ok(stream)) = incoming.next().await {
        tx.send(stream).await;
    }
}

```
`Handler.start()`
```rust
pub async fn start(&self) {
    while let Some(stream) = self.receiver.recv().await {
        task::spawn(async move {
            loop {
                // Handle stream until connection closes
            }
        });
    }
}
```

It's a lot of recursion, but I have other fields in `Handler`, and this is the only way I found to be able to pass them to the actual stream handler

I can find other ways to do this, but I wanted to understand why it doesn't work.

BTW, I forgot to say: this worked fine in `async-std` 1.5

**Comment by jimblandy** [comment]
The code in the initial comment works for me - it prints "Receiving connections" and then hangs.

If you're using something like `signal-hook` that turns Ctrl+C into something well-behaved like a future to wait on, you can use the `race` method on a future to let it race with another future and return whichever one finishes first. (You'll need to enable the "unstable" feature.) That is, calling `connection_handler` returns some future `F`, and you have another future `G` that is ready when a Ctrl+C arrives, so you say:

    F.race(G).await;

**Comment by forkbomb9** [comment]
Now it works. Weird.
I _think_ it didn't work on async-std 1.6.0, as it stopped working with the switch to `smol`, but now with 1.7 it does work.

The tip about `race` is very useful, thanks!
